query.go

```diff
@@ -947,75 +949,96 @@ func (s *ChainService) GetBlock(blockHash chainhash.Hash,
-	// which is always called single-threadedly. We don't check the block
-	// until after the query is finished, so we can just write to it
-	// naively.
 	var foundBlock *btcutil.Block
-	s.queryPeers(
-		// Send a wire.GetDataMsg
-		getData,
-
-		// Check responses and if we get one that matches, end the
-		// query early.
-		func(sp *ServerPeer, resp wire.Message,
-			quit chan<- struct{}) {
-			switch response := resp.(type) {
-			// We're only interested in "block" messages.
-			case *wire.MsgBlock:
-				// Only keep this going if we haven't already
-				// found a block, or we risk closing an already
-				// closed channel.
-				if foundBlock != nil {
-					return
+	request := &query.Request{
+		Req: getData,
+		HandleResp: func(req, resp wire.Message, peer string) query.Progress {
+			// The request must have been a "getdata" msg.
+			_, ok := req.(*wire.MsgGetData)
+			if !ok {
+				return query.Progress{
+					Finished:   false,
+					Progressed: false,
 				}
+			}
 
-				// If this isn't our block, ignore it.
-				if response.BlockHash() != blockHash {
-					return
+			// We're only interested in "block" messages.
+			response, ok := resp.(*wire.MsgBlock)
+			if !ok {
+				return query.Progress{
+					Finished:   false,
+					Progressed: false,
 				}
-				block := btcutil.NewBlock(response)
+			}
 
-				// Only set height if btcutil hasn't
-				// automagically put one in.
-				if block.Height() == btcutil.BlockHeightUnknown {
-					block.SetHeight(int32(height))
+			// If this isn't the block we asked for, ignore it.
+			if response.BlockHash() != blockHash {
+				return query.Progress{
+					Finished:   false,
+					Progressed: false,
 				}
+			}
 
-				// If this claims our block but doesn't pass
-				// the sanity check, the peer is trying to
-				// bamboozle us. Disconnect it.
-				if err := blockchain.CheckBlockSanity(
-					block,
-					// We don't need to check PoW because
-					// by the time we get here, it's been
-					// checked during header
-					// synchronization
-					s.chainParams.PowLimit,
-					s.timeSource,
-				); err != nil {
-					log.Warnf("Invalid block for %s "+
-						"received from %s -- "+
-						"disconnecting peer", blockHash,
-						sp.Addr())
-					sp.Disconnect()
-					return
+			block := btcutil.NewBlock(response)
+
+			// Only set height if btcutil hasn't
+			// automagically put one in.
+			if block.Height() == btcutil.BlockHeightUnknown {
+				block.SetHeight(int32(height))
+			}
+
+			// If this claims our block but doesn't pass
+			// the sanity check, the peer is trying to
+			// bamboozle us.
+			if err := blockchain.CheckBlockSanity(
+				block,
+				// We don't need to check PoW because
+				// by the time we get here, it's been
+				// checked during header
+				// synchronization
+				s.chainParams.PowLimit,
+				s.timeSource,
+			); err != nil {
+				log.Warnf("Invalid block for %s "+
+					"received from %s -- ",
+					blockHash, peer)
+				fmt.Println(err)
```

Comment:
Prior logic would disconnect here, but now we'll continue... 

Seems worthy of a future spin off to propagate a "ban worthy" error back up to the main scheduler.